controller/controller.go

```diff
@@ -0,0 +1,124 @@
+package controller
+
+import (
+	"compress/gzip"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/brave/go-sync/auth"
+	"github.com/brave/go-sync/command"
+	"github.com/brave/go-sync/datastore"
+	"github.com/brave/go-sync/sync_pb"
+	"github.com/brave/go-sync/timestamp"
+	"github.com/go-chi/chi"
+	"github.com/golang/protobuf/proto"
+	"github.com/rs/zerolog/log"
+)
+
+// SyncRouter add routers for command and auth endpoint requests.
+func SyncRouter(datastore datastore.Datastore) chi.Router {
+	r := chi.NewRouter()
+	r.Post("/command/", Command(datastore))
+	r.Post("/auth", Auth(datastore))
+	r.Get("/timestamp", Timestamp)
+	return r
+}
+
+func sendJSONRsp(body []byte, w http.ResponseWriter) {
+	w.Header().Set("Content-Type", "application/json")
+	_, err := w.Write(body)
+	if err != nil {
+		log.Error().Err(err).Msg("Write HTTP response body failed")
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+}
+
+// Timestamp returns a current timestamp back to sync clients.
+func Timestamp(w http.ResponseWriter, r *http.Request) {
+	body, err := timestamp.GetTimestamp()
+	if err != nil {
+		log.Error().Err(err).Msg("Get timestamp failed")
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+	sendJSONRsp(body, w)
+}
+
+// Auth handles authentication requests from sync clients.
+func Auth(db datastore.Datastore) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		_, body, err := auth.Authenticate(r, db)
+		if err != nil {
+			log.Error().Err(err).Msg("Authenticate failed")
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		sendJSONRsp(body, w)
+	})
+}
+
+// Command handles GetUpdates and Commit requests from sync clients.
+func Command(db datastore.Datastore) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// Authorize
+		clientID, err := auth.Authorize(db, r)
+		if clientID == "" {
+			if err != nil {
+				log.Error().Err(err).Msg("Authorization failed")
+			}
+			http.Error(w, http.StatusText(http.StatusUnauthorized), http.StatusUnauthorized)
+			return
+		}
+
+		// Decompress
+		var msg []byte
+		var gr *gzip.Reader
+		if r.Header.Get("Content-Encoding") == "gzip" {
+			gr, err = gzip.NewReader(r.Body)
+			if err != nil {
+				log.Error().Err(err).Msg("Gzip decompression failed")
+				http.Error(w, "Decompress error", http.StatusInternalServerError)
+				return
+			}
+			msg, err = ioutil.ReadAll(gr)
+		} else {
+			msg, err = ioutil.ReadAll(r.Body)
+		}
+		if err != nil {
+			log.Error().Err(err).Msg("Read request body failed")
+			http.Error(w, "Decompress error", http.StatusInternalServerError)
+			return
+		}
+
+		// Unmarshal into ClientToServerMessage
+		pb := &sync_pb.ClientToServerMessage{}
+		err = proto.Unmarshal(msg, pb)
```

Comment:
Its considered not super great form to use ioutil.ReadAll for a request body.  I would consider looking at:
github.com/gogo/protobuf/io
It seems like you can do something like this:
```go
io.NewFullReader(r.Body, maxSize).ReadMsg(pb)
```
where 1024 is the max size to read.  This would save you from using unmarshal and doing readall.

Reply:
Addressed in https://github.com/brave/go-sync/pull/1/commits/626c63b964e4181f6e8648319d00810f6a615ad6